aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -319,78 +319,234 @@ bool is_safe_to_get_storage_as_tensor(const NestedTensorImpl* tensor) {
+  const int64_t q_batch_size = query.size(0);
+  const int64_t k_batch_size = key.size(0);
+  const int64_t v_batch_size = value.size(0);
 
-  // Query -> Query (Batch x {Q_seq_len}  x Num_heads x Dim_per_head)
-  // Key   -> Key   (Batch x {KV_seq_len} x Num_heads x Dim_per_head)
-  // Value -> Value (Batch x {KV_seq_len} x Num_heads x Dim_per_head)
-  Tensor q_t = query.transpose(1, 2).contiguous();
-  Tensor k_t = key.transpose(1, 2).contiguous();
-  Tensor v_t = value.transpose(1, 2).contiguous();
+  const int64_t output_batch_size = std::max({q_batch_size, k_batch_size, v_batch_size});
 
-  // K and V have to have the same Nnz, should probably torch_check
-  // assume in order to not iterate over v
+  const int64_t q_num_heads = query.size(1);
+  const int64_t k_num_heads = key.size(1);
+  const int64_t v_num_heads = value.size(1);
+
+  const int64_t output_num_heads = std::max({q_num_heads, k_num_heads, v_num_heads});
+
+  const int64_t head_dim_qk = query.size(3);
+  const int64_t head_dim_v = value.size(3);
+
+  Tensor q_t = query.transpose(1, 2);
+  Tensor k_t = key.transpose(1, 2);
+  Tensor v_t = value.transpose(1, 2);
 
-  auto cumulative_and_max_q = cumulative_and_max_seq_len(q_t);
-  auto cumulative_and_max_k = cumulative_and_max_seq_len(k_t);
+  // Checks in sdp_utils ensure that if {*}_batch_size/{*}_num_heads != output_batch_size/num_heads
+  // then they are 1
+  bool q_batch_size_needs_broadcast = q_batch_size != output_batch_size;
+  bool k_batch_size_needs_broadcast = k_batch_size != output_batch_size;
+  bool v_batch_size_needs_broadcast = v_batch_size != output_batch_size;
+
+  int64_t max_seqlen_batch_q = 0, Nnz_q = 0;
+  Tensor cumulative_sequence_length_q;
+  // initialize Nnz as well
+  if (q_batch_size_needs_broadcast || !q_t.is_nested()) {
+    max_seqlen_batch_q = q_t.size(1);
+    cumulative_sequence_length_q = at::arange(0,
+                                              (output_batch_size + 1) * max_seqlen_batch_q,
+                                              max_seqlen_batch_q,
+                                              TensorOptions().device(at::kCUDA).dtype(at::kInt));
+    Nnz_q = output_batch_size * max_seqlen_batch_q;
+  } else {
+    auto cumulative_and_max_q_and_nnz_q = cumulative_and_max_seq_len(q_t);
+    cumulative_sequence_length_q = std::get<0>(cumulative_and_max_q_and_nnz_q);
+    max_seqlen_batch_q = std::get<1>(cumulative_and_max_q_and_nnz_q);
+    Nnz_q = std::get<2>(cumulative_and_max_q_and_nnz_q);
+  }
 
-  Tensor cumulative_sequence_length_q = std::get<0>(cumulative_and_max_q);
-  Tensor cumulative_sequence_length_k = std::get<0>(cumulative_and_max_k);
+  int64_t max_seqlen_batch_kv = 0, Nnz_kv = 0;
+  Tensor cumulative_sequence_length_kv;
+  if (k_batch_size_needs_broadcast) {
+    if (v_batch_size_needs_broadcast) {
+      // Should not be the case that both k and v need broadcasting of batch_size
+      // i.e. output_batch_size == 1 and k/v_batch_size are both 1
```

Comment:
!= 1 

Also is this checked in sdp_utils?

Reply:
Will add a check, thanks for catching

Reply:
edit: actually @drisspg, do you think this case output_batch_size == 1 and k/vbatch_size == 1 might potentially sense? rethinking it now I I might have missed it here cuz then it would just be

k, v `[1, num_head, x, embed_qk], [1, num_head, x, embed_v] --> [B, num_head, x, embed_qk], [B, num_head, x, embed_v]`

The case I actually missed was if 1 of k/v batch_size needs broadcast and not the other then the other must have a consistent seq_len dim, e.g. 

k, v `[1, num_head, x, embed_qk], [B, num_head, x, embed_v] --> [B, num_head, x, embed_qk], [B, num_head, x, embed_v]`

does that sound right to you?

Reply:
To be honest this is the most logic dense part of the code and little hard to follow all the paths. But the broadcasting examples above I agree with. Like you said in the second example when k and v are nested we need to be more careful on the seq_len broadcasting